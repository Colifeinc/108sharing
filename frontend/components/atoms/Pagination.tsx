import range from 'lodash-es/range'
import * as React from 'react'
import classNames from 'classnames'
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons/faChevronLeft'
import { faChevronRight } from '@fortawesome/free-solid-svg-icons/faChevronRight'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

interface IProps {
  currentPage: number
  prevPage: number | null
  nextPage: number | null
  totalPages: number
  totalCount: number
  onChangePageHandler?(page: number): void
  className?: string
}

interface IPaginationItem {
  content: number | string | React.ReactElement
  onClick?: () => void
  disabled?: boolean
  active?: boolean
  isIcon?: boolean
}

const PAGE_LIMIT = 3

const Pagination: React.FC<IProps> = ({
  currentPage,
  prevPage,
  nextPage,
  totalPages,
  totalCount,
  onChangePageHandler,
  className = '',
}) => {
  const onChangePage = React.useCallback(
    (page: number) => {
      if (currentPage === page) return
      onChangePageHandler(page)
    },
    [currentPage, onChangePageHandler]
  )

  if (totalPages <= 1) {
    return null
  }

  const pages = range(currentPage - 1, currentPage - 1 + PAGE_LIMIT).filter(
    page => page > 1 && page < totalPages
  )
  const items: IPaginationItem[] = [
    {
      content: <FontAwesomeIcon icon={faChevronLeft} />,
      onClick: !prevPage ? null : () => onChangePage(currentPage - 1),
      disabled: !prevPage,
      active: false,
      isIcon: true,
    },
    {
      content: 1,
      onClick: () => (currentPage === 1 ? null : onChangePage(1)),
      active: currentPage === 1,
    },
    pages[0] > 2 && {
      content: '...',
    },
    ...pages.map(page => {
      return {
        content: page,
        onClick: currentPage === page ? null : () => onChangePage(page),
        active: currentPage === page,
      }
    }),
    pages[pages.length - 1] + 1 < totalPages && {
      content: '...',
    },
    {
      content: totalPages,
      onClick: currentPage === totalPages ? null : () => onChangePage(totalPages),
      active: currentPage === totalPages,
    },
    {
      content: <FontAwesomeIcon icon={faChevronRight} />,
      onClick: !nextPage ? null : () => onChangePage(currentPage + 1),
      disabled: !nextPage,
      active: false,
      isIcon: true,
    },
  ].filter(Boolean)

  return (
    <nav className={classNames(['flex items-center justify-center font-bold', className])}>
      {items.map((item, i) => {
        const Tag = item.onClick ? 'a' : 'div'
        return (
          <Tag
            key={i}
            className={classNames([
              'w-8 h-8 flex items-center justify-center mx-2 rounded-full',
              item.onClick && 'cursor-pointer transition duration-150 hover:opacity-85',
              !item.isIcon && item.active && 'text-white bg-primary',
              !item.isIcon && !item.active && 'text-primary',
              item.isIcon && item.disabled && 'text-neutral-500',
              item.isIcon && !item.disabled && 'text-primary shadow-icon',
            ])}
            onClick={item.onClick}
          >
            <div>{item.content}</div>
          </Tag>
        )
      })}
    </nav>
  )
}

export default Pagination
