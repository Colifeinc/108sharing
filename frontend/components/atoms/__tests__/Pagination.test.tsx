import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import Pagination from '../Pagination'

describe('atoms/Pagination.tsx', () => {
  const onChangePageHandler = jest.fn()
  const props = {
    currentPage: 3,
    prevPage: 2,
    nextPage: 4,
    totalPages: 5,
    totalCount: 100,
    onChangePageHandler,
  }

  it('should matches to the snapshot', () => {
    const wrapper = shallow(<Pagination {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  // it('should call onChangePageHandler with next page', () => {
  //   const wrapper = shallow(<Pagination {...props} />)
  //   wrapper.find('.Pagination_Arrow._next').simulate('click')
  //   expect(onChangePageHandler).toHaveBeenCalledWith(4)
  // })

  // it('should call onChangePageHandler with prev page', () => {
  //   const wrapper = shallow(<Pagination {...props} />)
  //   wrapper.find('.Pagination_Arrow._prev').simulate('click')
  //   expect(onChangePageHandler).toHaveBeenCalledWith(2)
  // })
})
