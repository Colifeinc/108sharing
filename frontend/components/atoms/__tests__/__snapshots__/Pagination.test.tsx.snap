// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Pagination.tsx should matches to the snapshot 1`] = `
<nav
  className="flex items-center justify-center font-bold"
>
  <a
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full cursor-pointer transition duration-150 hover:opacity-85 text-primary shadow-icon"
    key="0"
    onClick={[Function]}
  >
    <div>
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              320,
              512,
              Array [],
              "f053",
              "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
            ],
            "iconName": "chevron-left",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </div>
  </a>
  <a
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full cursor-pointer transition duration-150 hover:opacity-85 text-primary"
    key="1"
    onClick={[Function]}
  >
    <div>
      1
    </div>
  </a>
  <a
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full cursor-pointer transition duration-150 hover:opacity-85 text-primary"
    key="2"
    onClick={[Function]}
  >
    <div>
      2
    </div>
  </a>
  <div
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full text-white bg-primary"
    key="3"
    onClick={null}
  >
    <div>
      3
    </div>
  </div>
  <a
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full cursor-pointer transition duration-150 hover:opacity-85 text-primary"
    key="4"
    onClick={[Function]}
  >
    <div>
      4
    </div>
  </a>
  <a
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full cursor-pointer transition duration-150 hover:opacity-85 text-primary"
    key="5"
    onClick={[Function]}
  >
    <div>
      5
    </div>
  </a>
  <a
    className="w-8 h-8 flex items-center justify-center mx-2 rounded-full cursor-pointer transition duration-150 hover:opacity-85 text-primary shadow-icon"
    key="6"
    onClick={[Function]}
  >
    <div>
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              320,
              512,
              Array [],
              "f054",
              "M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z",
            ],
            "iconName": "chevron-right",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </div>
  </a>
</nav>
`;
